consensus/src/fallback.rs

```diff
@@ -0,0 +1,770 @@
+            let mut next_leader = self.leader_elector.get_leader(self.round + 1);
+            // For fallback blocks, send vote back to the proposer.
+            if block.fallback == 1 {
+                next_leader = block.author;
+            }
+            if next_leader == self.name {
+                self.handle_vote(&vote).await?;
+            } else {
+                let message = CoreMessage::Vote(vote);
+                self.transmit(&message, Some(next_leader)).await?;
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn handle_proposal(&mut self, block: &Block) -> ConsensusResult<()> {
+        let digest = block.digest();
+
+        // Ensure the block proposer is the right leader for the round.
+        ensure!(
+            block.fallback == 1 || block.author == self.leader_elector.get_leader(block.round),
+            ConsensusError::WrongLeader {
+                digest,
+                leader: block.author,
+                round: block.round
+            }
+        );
+
+        // Check the block is correctly formed.
+        block.verify(&self.committee)?;
+
+        // Process the QC. This may allow us to advance round.
+        self.process_qc(&block.qc).await;
+
+        // // Process the TC (if any). This may also allow us to advance view.
+        // if let Some(ref tc) = block.tc {
+        //     tc.verify(&self.committee)?;
+        //     self.advance_view(tc.seq).await;
+        // }
+
+        // Let's see if we have the block's data. If we don't, the mempool
+        // will get it and then make us resume processing this block.
+        if !self.mempool_driver.verify(block).await? {
+            debug!("Processing of {} suspended: missing payload", digest);
+            return Ok(());
+        }
+
+        // Not in fallback, process the fallback blocks when later enter the fallback
+        // It is necessary to receive 2f+1 timeout messages with QCs to update the high_qc
+        if block.fallback == 1 && self.fallback == 0 {
+            if block.height == 1 || block.height == 2 {
+                let map = self.fallback_pending_blocks.entry(block.view).or_insert(HashMap::new());
+                map.insert((block.author, block.height), block.clone());
+            }
+            return Ok(());
+        }
+
+        // All check pass, we can process this block.
+        self.process_block(block).await
+    }
+
+    async fn handle_sync_request(
+        &mut self,
+        digest: Digest,
+        sender: PublicKey,
+    ) -> ConsensusResult<()> {
+        if let Some(bytes) = self.store.read(digest.to_vec()).await? {
+            let block = bincode::deserialize(&bytes)?;
+            let message = CoreMessage::Propose(block);
+            self.transmit(&message, Some(sender)).await?;
+        }
+        Ok(())
+    }
+
+    // With async fallback, do not handle TC directly. Need to receive 2f+1 Timeout to update the high QC.
+    async fn handle_tc(&mut self, _: TC) -> ConsensusResult<()> {
```

Comment:
Is this a TODO?

Reply:
No, just explaining why we cannot handle TC directly